apps/prairielearn/src/middlewares/clientFingerprint.ts

```diff
@@ -0,0 +1,46 @@
+import asyncHandler = require('express-async-handler');
+import * as sqldb from '@prairielearn/postgres';
+import { IdSchema } from '../lib/db-types';
+
+const sql = sqldb.loadSqlEquiv(__filename);
+
+export default asyncHandler(async (req, res, next) => {
+  const user_session_id = await sqldb.queryOptionalRow(
+    sql.select_session_id,
+    { user_session_id: req.session.id },
+    IdSchema,
+  );
+
+  const params = {
+    ip_address: req.ip,
+    user_id: res.locals.authn_user.user_id,
+    user_session_id: user_session_id,
+    user_agent: req.headers['user-agent'],
+    accept_language: req.headers['accept-language'],
+  };
+  let client_fingerprint_id = await sqldb.queryOptionalRow(
+    sql.select_client_fingerprint,
+    params,
+    IdSchema,
+  );
+
+  if (!client_fingerprint_id) {
+    client_fingerprint_id = await sqldb.queryRow(sql.insert_client_fingerprint, params, IdSchema);
+  }
+  if (res.locals.assessment_instance) {
```

Comment:
Yes, this is currently only mounted on routes that have an assessment instance. Originally I thought I needed to explicitly state that `res.locals.assessment_instance` needed to exist in order to comply with TypeScript. Though it looks like that may not be the case. Would you like me to remove this line?

Reply:
Yeah, I think this is already as efficient as it can be. My preference would be to make the check for `res.locals.assessment_instance` the first thing that happens in this middleware, and throw an error if it isn't present. I'd rather fail loudly here if we end up putting this middleware somewhere it doesn't belong.

Reply:
Yep, I agree that's a good pattern here.

Reply:
I have moved the check for `res.locals.assessment_instance` to the top of of the middleware and throws a new error if the assessment_instance does not exist. This is change is made in commit [c4eb708](https://github.com/PrairieLearn/PrairieLearn/pull/8798/commits/c4eb70857d0fce8cd75fdefd6e3694c0850b25b7).